### PR TITLE
Extend Python API (rand, lm, matrix multiplication)

### PR DESCRIPTION
**Pull request by juliale-15**
Add rand(), lm and matrix multiplication to Python API

### CODE REVIEW

File: src/main/python/systemds/matrix/operation_node.py [hunk 1]
```diff
@@ -229,3 +233,10 @@ def moment(self, moment, weights: DAGNode = None) -> 'OperationNode':
             unnamed_inputs.append(weights)
         unnamed_inputs.append(moment)
         return OperationNode(self.sds_context, 'moment', unnamed_inputs, output_type=OutputType.DOUBLE)
+
+    def lm(self, y: DAGNode, **kwargs) -> 'OperationNode':
```
Comment (Baunsgaard) [inline src/main/python/systemds/matrix/operation_node.py:237]: Missing tests (You said this, but just a reminder.)

File: src/main/python/systemds/matrix/operation_node.py [hunk 2]
```diff
@@ -159,6 +159,10 @@ def __eq__(self, other):
     def __ne__(self, other):
         return OperationNode(self.sds_context, '!=', [self, other])
 
+
```
Comment (Baunsgaard) [inline src/main/python/systemds/matrix/operation_node.py:162]: (OCD comment) double newline

File: src/main/python/systemds/matrix/operation_node.py (hunk 1 above)
Comment (Baunsgaard) [inline src/main/python/systemds/matrix/operation_node.py:237]: (Suggestion) Could it make sense to move this lm function to a new file?

I can imagine that if we add all functions to this file it will become harder to look over.

File: src/main/python/systemds/matrix/matrix.py [hunk 3]
```diff
@@ -141,3 +141,32 @@ def seq(sds_context: 'SystemDSContext', start: Union[float, int], stop: Union[fl
         start = 0
     unnamed_input_nodes = [start, stop, step]
     return OperationNode(sds_context, 'seq', unnamed_input_nodes)
+
+
```
Comment (Baunsgaard) [inline src/main/python/systemds/matrix/matrix.py:145]: (OCD comment) double newline

File: src/main/python/tests/test_matrix_aggregations.py [hunk 4]
```diff
@@ -84,6 +86,63 @@ def test_var2(self):
     def test_var3(self):
         self.assertTrue(np.allclose(sds.matrix(m1).var(axis=1).compute(), m1.var(axis=1, ddof=1).reshape(dim, 1)))
 
+    def test_rand_basic(self):
+        seed = 15
+        shape = (20, 20)
```
[the remark this answers is not in the document]
Comment (juliale-15) [reply src/main/python/tests/test_matrix_aggregations.py:91]: I would expect an exception to be thrown, and the user can catch it in a try-except block. 
In this case I would add a test case to expect an exception in a try except block.
Is this what you meant?

File: src/main/python/tests/test_matrix_aggregations.py [hunk 5]
```diff
@@ -25,6 +25,8 @@
 import warnings
 import unittest
 import numpy as np
+import scipy.stats as st
```
[the remark this answers is not in the document]
Comment (juliale-15) [reply src/main/python/tests/test_matrix_aggregations.py:28]: Yes. I think it would be good to add it :) it can probably be used for other testcases

File: src/main/python/tests/test_matrix_aggregations.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Baunsgaard) [reply src/main/python/tests/test_matrix_aggregations.py:91]: yes, but while doing it, look at the exception thrown, and the output as an entire thing, and reason with yourself, does this error message help. If yes, GREAT! If not what changes to that error message would be good to help a user, and where in the operations should this information be extracted and returned to the user.

File: src/main/python/tests/test_matrix_aggregations.py [hunk 6]
```diff
@@ -84,6 +86,63 @@ def test_var2(self):
...
+
+    for dist in distributions:
+        param = getattr(st, dist).fit(data)
+
+        D, p_value = st.kstest(data, dist, args=param)
+        result[dist] = p_value
+
+    best_dist = max(result, key=result.get)
+    return best_dist
+
```
[the remark this answers is not in the document]
Comment (juliale-15) [reply src/main/python/tests/test_matrix_aggregations.py:145]: Hi, regarding the code formatter,  I'm using PyCharms integrated Pep8 checks. Would that be enough?

[the remark this answers is not in the document]
Comment (Baunsgaard) [reply src/main/python/tests/test_matrix_aggregations.py:145]: if it removes the newlines, that would be great :+1:

File: src/main/python/systemds/context/systemds_context.py [hunk 7]
```diff
@@ -164,4 +164,14 @@ def rand(self, rows: int, cols: int, min: Union[float, int] = None,
         :param lambd: lamda value for "poison" distribution
         :return:
         """
+        available_pdfs = ["uniform", "normal", "poisson"]
+        if rows < 0:
+            raise ValueError("In rand statement, can only assign rows a long (integer) value >= 0 "
+                            "-- attempted to assign value: {r}".format(r=rows))
+        if cols < 0:
+            raise ValueError("In rand statement, can only assign cols a long (integer) value >= 0 "
```
Comment (Baunsgaard) [inline src/main/python/systemds/context/systemds_context.py:172]: `if cols <= 0:`
`In rand statement, can only assign cols a long (integer) value > 0 `

not >=0

This makes me think that you did not verify the edge case of 0. Other than that, the message looks good :smile:

File: src/main/python/tests/test_matrix_rand.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    for dist in distribution_lst:
+        param = getattr(st, dist).fit(data)
+
+        D, p_value = st.kstest(data, dist, args=param)
+        result[dist] = p_value
+
+    best_dist = max(result, key=result.get)
+    return best_dist
+
+
```
Comment (Baunsgaard) [inline src/main/python/tests/test_matrix_rand.py:108]: OCD (double newline)

File: src/main/python/tests/test_matrix_rand.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+            print(e)
+
+    def test_rand_invalid_pdf(self):
+        try:
+            sds.rand(rows=1, cols=10, pdf="norm").compute()
+            self.assertTrue(False)
+        except Exception as e:
+            print(e)
+
+
```
Comment (Baunsgaard) [inline src/main/python/tests/test_matrix_rand.py:89]: OCD (double newline)

File: src/main/python/tests/test_matrix_rand.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+shape = (random.randrange(0, 50), random.randrange(0, 50))
+min_max = (0, 1)
+sparsity = 0.2
+distributions = ['norm', 'uniform']
+
+sds = SystemDSContext()
+
+
```
Comment (Baunsgaard) [inline src/main/python/tests/test_matrix_rand.py:42]: OCD (double newline)

File: src/main/python/tests/test_matrix_rand.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+import unittest
+import numpy as np
+import scipy.stats as st
+import random
+
+path = os.path.join(os.path.dirname(os.path.realpath(__file__)), "../")
+sys.path.insert(0, path)
+from systemds.context import SystemDSContext
+
+
```
Comment (Baunsgaard) [inline src/main/python/tests/test_matrix_rand.py:34]: OCD (double newline)

File: src/main/python/systemds/context/systemds_context.py (hunk 7 above)
Comment (juliale-15) [reply src/main/python/systemds/context/systemds_context.py:172]: I did not include the zero case because `np.random.randint(0,1,(0,0))` is a valid call. And `sds.rand(0,0)` returns the same output (an empty numpy array). Should I change the condition anyways?

Comment (Baunsgaard) [reply src/main/python/systemds/context/systemds_context.py:172]: hmm, then duplicate the behavior from NP, and see if it creates an empty output. How does DASK do it? also empty array?

**Comment by j143** [comment]
Hi, How is this going?

**Comment by j143** [comment]
@Baunsgaard - Does this PR look good to you.
Suggestion: A more involved test can be handled in a different PR. 😄

File: src/main/python/systemds/matrix/operation_node.py (hunk 1 above)
Comment (juliale-15) [reply src/main/python/systemds/matrix/operation_node.py:236]: Hi, do you have any suggestion on how the file should be called?

Comment (Baunsgaard) [reply src/main/python/systemds/matrix/operation_node.py:236]: Something along the lines of "buildin", and if you make it then it would be great to move l2svm along with it.
But, lest do that in another PR at some later point, because then we will merge this PR shortly

File: src/main/python/systemds/context/systemds_context.py (hunk 7 above)
Comment (Baunsgaard) [reply src/main/python/systemds/context/systemds_context.py:172]: Just to finish this thread, you have done the correct thing! :100:, I did not consider that it is useful to make code that have a 0x0 matrix to start with, and then add columns and rows to that, iteratively.

**Comment by Baunsgaard** [comment]
LGTM, Many thanks @juliale-15 :100:.
Absolutely great first addition! I will start using them immediately, hope you learned something! :smile: 

@j143  Yes it can be merged as it is, the component test fail because of other reasons.

**Comment by mboehm7** [comment]
Great, thank you so much @juliale-15 and @Baunsgaard - I'll merge it in.

**Comment by juliale-15** [comment]
> Great, thank you so much @juliale-15 and @Baunsgaard - I'll merge it in.

Great thank you! Yesterday I did a small fix in the rand and lm test and didn't realized that this PR is already closed. Should I just add the changes to the next PR? I had to increase the tolerance in the testcases so that it surely never fails.

**Comment by Baunsgaard** [comment]
Yes, make a new PR :smile:, sorry for the spontaneous merge :upside_down_face:.
